### ISSUE TITLE
Update dependency before destroy when no replacement created

### ISSUE THREAD

**Issue by TyeMcQueen** [hashicorp/terraform#32136]
### Terraform Version

```shell
Terraform v1.3.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/google v4.42.0
```

### Terraform Configuration Files

```terraform

variable "step" {
  description   = "0, 1, 2, or 3"
  type          = number
}

resource "google_project_service" "required" {
  for_each  = toset([
    "certificatemanager.googleapis.com",
  ])
  service   = each.value

  disable_on_destroy            = false
  disable_dependent_services    = false
}

resource "google_certificate_manager_certificate" "c1" {
  count         = 2 == var.step ? 1 : 0
  name          = "cert1"
  managed {
    domains     = [ "host1.example.com" ]
  }
# lifecycle { create_before_destroy = true }
}

resource "google_certificate_manager_certificate" "c2" {
  count         = 0 < var.step ? 1 : 0
  name          = "cert2"
  managed {
    domains     = [ "host1.example.com" ]
  }
}

resource "google_certificate_manager_certificate_map" "m" {
  count         = 0 < var.step ? 1 : 0
  name          = "map"
}

resource "google_certificate_manager_certificate_map_entry" "e" {
  count         = var.step < 2 ? 0 : 1
  map           = google_certificate_manager_certificate_map.m[0].name
  name          = "ent"
  certificates  = [ 3 == var.step
    ? google_certificate_manager_certificate.c2[0].id
    : google_certificate_manager_certificate.c1[0].id ]
  hostname      = "host1.example.com"
# lifecycle { create_before_destroy = true }
}

```

### Debug Output

https://gist.github.com/TyeMcQueen/bafd9fce21973a1686c0bab6215b944e

### Expected Behavior

The dependency relationship between c1 and ent should cause ent to be updated before c1 is deleted.  This would lead to the apply always succeeding.

### Actual Behavior

At least sometimes, the first action is to attempt to delete c1 which fails because c1 is being used by ent.  Once this happens, trying to `apply` again will fail in the same way.

### Steps to Reproduce

1. Config auth to a GCP project
2. terraform init
3. terraform apply -var step=0 -auto-approve
4. terraform apply -var step=2 -auto-approve && terraform apply -var step=3 -auto-approve
5. This will likely produce the documented failure.  If you get success, then go back to Step 3.
6. If desired, uncomment the "create_before_destroy" lines to show they don't fix the problem
7. terraform apply -var step=3 -auto-approve

### Additional Context

Searching, I found forms of this bug previously reported more than once.  However, the resolutions always got focused on specific details which derailed properly identifying a real bug in base Terraform that should just be fixed.  The specifics provided above are just one of the simplest demonstrations of the problem I could come up with but this problem impacts many resource types (any resources that enforce referential integrity).

Some background on how Terraform works to help frame the problem (because this came up repeatedly when this problem was previously reported)...

Terraform has 2 designed behaviors that are related to ordering of updates vs deletes.  But these only cover the case of the child being recreated.  That is, if parent references child and we need to recreate child and update parent to now point to the new child, if `lifecycle { create_before_destroy = true }` is set for child, we get:

1. create child v2
2. update parent to move from child v1 to child v2
3. delete child v1

This is the best approach when constraints don't prevent having 2 versions of child simultaneous (which often gets worked around by adding a random suffix to child's name).

If you don't set `create_before_destroy`, then Terraform presumes that it isn't safe to try to create 2 simultaneous instances of child so delete must happen before create and we get:

1. delete child v1
2. create child v2
3. update parent to move to child v2

Step 1 can fail here due to referential integrity checks.

But adding `create_before_destroy` is not a solution after you get stuck with the problem being reported.  Oddly, setting it before you have the problem appears to prevent the problem from happening (or at least make it less likely).

THE PROBLEM

Terraform lacks (and needs) a rule for ordering of updates vs deletes when the delete is not part of a recreate.

If parent references child and the plan calls for parent to be changed to no longer reference child and to also simply delete child, then Terraform should always do the update before the delete:

1. update parent to no longer reference child
2. delete child

Experimentation shows that the order here may not always be consistent even in the most recent version of Terraform.  It isn't random to the point of just being able to do multiple `terraform apply`s until it works, however. 

I believe it would be sufficient to not have to care exactly what the update operation on parent is.  Just if parent is being updated, child is being deleted, and parent references child, then update before you delete.

Yes, there are troublesome ways to work around this problem (apply the change in 2 steps or use -target).  But it would be very beneficial for Terraform to just handle doing these operations in the correct order and not sometimes require by-hand wrangling to work your way out of the this situations when it crops up.

Even if `create_before_destroy` is always sufficient to prevent the problem if you do it before you have the problem (my experimentation suggests that it might be but I am not certain), that is not the best solution.  There are plenty of situations where "update before delete" is just the correct order yet `create_before_destroy` is not safe.  In fact, if you managed to trigger a "recreate" of c1 here, it would fail if you had `create_before_destroy` set.

**Comment by jbardin** [comment]
Hi @TyeMcQueen,

There are a few limitations which are going to prevent making a change like this. 

 - Terraform has a well-defined order if operations with resource dependencies. Existing providers must be able to trust that overall order is always maintained, regardless of the number of individual operations being applied. Because one example with this provider may be indifferent to a change in the order, it does not mean all examples with all providers are indifferent to changing the order -- e.g. it is perfectly valid to expect that the update to `e` cannot happen until after `c1` has been deleted.
 - The order of operations must be consistent for users. If the order were conditionally changed here, it would mean the update result could be different depending on whether the dependency change is a replace or delete, which would be incorrect for Terraform and similarly confusing for users.
 - All proposed graph transformations must consistently result in a valid DAG. Terraform needs an acyclic graph for ensuring that all operations can be completed. Reversing this edge with an upstream dependency being changed results in a dependency cycle. Conditionally changing this order only when it doesn't result in a cycle brings us back to the second point, where it's inconsistent for the user, as there is not direct way to predict what the order will be. Silently switching the upstream order to `create_before_destroy` to prevent cycles when there is only an intermediary delete on one particular resource would be similarly confusing, and causes one of the same problems you were trying to avoid here.

Terraform has a well-defined order of operations, if you have found that the evaluation order between 2 dependencies is not consistent, that is a separate issue which needs to be addressed. 

You can see #31769 for another explanation with a slightly different example.

For a primer on the internal ordering of these operations, you can see some notes I made here: https://github.com/hashicorp/terraform/blob/main/docs/destroying.md.

Since Terraform is working as designed this does not represent a bug in terraform, so I'm going to close the issue. If you have more questions, feel free to start a thread in [the community forum](https://discuss.hashicorp.com/c/terraform-core) for more discussion.

Thanks!
